### ISSUE TITLE
"proto/x.hpp" not found on NixOS

### ISSUE THREAD

**Issue by mrkgnao** [jaagr/polybar#308]
Am I missing a library?

**Comment by jaagr** [comment]
Could you show the build log? The file is generated by the `xpp` lib.

**Comment by mrkgnao** [comment]
Yeah. I'm also including the list of dependencies I'm using.

[polybar.nix.txt](https://github.com/jaagr/polybar/files/692211/polybar.nix.txt)
[build.txt](https://github.com/jaagr/polybar/files/692210/build.txt)

**Comment by jaagr** [comment]
What's the error message after the last line in build.txt? Is it the message about the missing header?

**Comment by mrkgnao** [comment]
Yes.

**Comment by jaagr** [comment]
This might be a cmake issue. I would guess that the headers generated by xpp isn't available in the include path. Could you post your `compile_commands.json`

**Comment by jaagr** [comment]
And `ls -1 lib/xpp/include/{,proto}`

**Comment by mrkgnao** [comment]
Here you go.

[build.txt](https://github.com/jaagr/polybar/files/692282/build.txt)
[compile_commands.json.txt](https://github.com/jaagr/polybar/files/692280/compile_commands.json.txt)
[ls.txt](https://github.com/jaagr/polybar/files/692281/ls.txt)

**Comment by jaagr** [comment]
The empty proto directory should contain the generated files. Will look at the build file for the xpp lib and see if I find something.

**Comment by mrkgnao** [comment]
Any workarounds, even if painful, in the meantime? I really need to get
this working.

**Comment by jaagr** [comment]
try:

~~~
cd lib/xpp
mkdir build
cmake ..
make
~~~

then rebuild the main project

**Comment by mrkgnao** [comment]
Well, it builds fine, but the main project still doesn't compile.

**Comment by jaagr** [comment]
Does it generate the proto files?

**Comment by jaagr** [comment]
You might need to recreate the main build folder

**Comment by mrkgnao** [comment]
No, it doesn't. I tried with a fresh clone.

I did `cd build` before the `cmake` step, though.

**Comment by jaagr** [comment]
Hm. Try installing python 3, I know it's required for arch to get the xcbgen python module

**Comment by jaagr** [comment]
Any luck?

**Comment by mrkgnao** [comment]
Hmm, removing python2 from the env and using python3 instead makes CMake complain about not being able to find `xcbgen`.

**Comment by jaagr** [comment]
python2 is still required, but it's not related to the xcbgen module anyway if that's what happened. Will see if I can find anything else

**Comment by corngood** [comment]
This doesn't look very nix friendly:
```
file(GLOB PROTO_LIST_RAW RELATIVE /usr/share/xcb /usr/share/xcb/*.xml)
```
I'll keep investigating.

**Comment by corngood** [comment]
https://github.com/jaagr/xpp/pull/4 gets me a bit further, but now I get:

```
[  2%] Building CXX object bin/CMakeFiles/polybar.dir/components/bar.cpp.o
<command-line>:0:1: error: macro names must be identifiers
```

It's probably something silly.  I haven't looked into it at all yet.

**Comment by jaagr** [comment]
What compiler/version?

**Comment by corngood** [comment]
gcc 5.4.0, on my modified version of NixOS 17.03.  I don't know if you're a nix user, but it has a pretty complex and fragile system for wrapping gcc, so it could be something wrong on the nix/nixpkgs side.

**Comment by jaagr** [comment]
The error is probably caused by the compiler flags:

~~~
 -D-I/nix/store/4ip9ag5mcg5zdgzvklgj16g3c12sfnim-libX11-1.6.4-dev/include
~~~

**Comment by corngood** [comment]
Yeah, I'm just trying to see where those are coming from.

**Comment by jaagr** [comment]
Try applying this patch

~~~patch
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 431a9d9..58f5164 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -137,11 +137,6 @@ make_executable(${PROJECT_NAME} SOURCES
 
 target_link_libraries(${PROJECT_NAME} Threads::Threads)
 
-target_compile_definitions(${PROJECT_NAME} PUBLIC
-  ${X11_Xft_DEFINITIONS}
-  ${X11_XCB_DEFINITIONS}
-  ${XCB_DEFINITIONS})
-
 # }}}
 # Create executable target: ipc messager {{{
 
~~~

**Comment by corngood** [comment]
Yeah that fixed it.  The first one didn't work on its own btw.

```
[100%] Linking CXX executable polybar
[100%] Built target polybar
```

Looks good.  Do you want me to do a PR, or have you got it?

Does something need to be fixed in the Find.cmake scripts?

**Comment by jaagr** [comment]
The fix should be applied to the libs malformed pkg-config file, because I would assume that the `Cflags` entry is wrong for them.

**Comment by corngood** [comment]
I think it's that it's using CFLAGS as _DEFINITIONS, thus prefixing with -D, e.g.
```
  SET(X11_XCB_DEFINITIONS ${PKG_X11_XCB_CFLAGS})
```

**Comment by jaagr** [comment]
Yeah that doesn't look right.

**Comment by jaagr** [comment]
Other Find modules seems to use `${PKG_xxx_CFLAGS_OTHER}` for compiler definitions

**Comment by corngood** [comment]
Yeah I see that for example in:

https://git.reviewboard.kde.org/r/117491/diff/2#index_header

I still think it should be added to cflags in cmake, not definitions.

On 12 Jan 2017 7:29 p.m., Michael Carlberg <notifications@github.com> wrote:

Other Find modules seems to use ${PKG_xxx_CFLAGS_OTHER} for compiler definitions

**Comment by jaagr** [comment]
Odd choice of variable names, but yeah I think it's safer to use `target_compile_options` instead, since most cmake modules seems to prepend the -D switch for the definitions anyway.

target_compile_definitions:
> This command can be used to add any flags, but it is intended to add preprocessor definitions...

target_compile_options:
> This command can be used to add any options, but alternative commands exist to add preprocessor definitions...

**Comment by corngood** [comment]
The current version of the cmake docs doesn't seem to say anything about `target_compile_definitions` being used for general flags, so maybe they changed that behaviour?

From the `FindPkgConfig` docs:

```
<XPREFIX>_FOUND          ... set to 1 if module(s) exist
<XPREFIX>_LIBRARIES      ... only the libraries (w/o the '-l')
<XPREFIX>_LIBRARY_DIRS   ... the paths of the libraries (w/o the '-L')
<XPREFIX>_LDFLAGS        ... all required linker flags
<XPREFIX>_LDFLAGS_OTHER  ... all other linker flags
<XPREFIX>_INCLUDE_DIRS   ... the '-I' preprocessor flags (w/o the '-I')
<XPREFIX>_CFLAGS         ... all required cflags
<XPREFIX>_CFLAGS_OTHER   ... the other compiler flags
```

So really FindXCB.cmake should just output those vars directly.  I have no idea why it's doing:

```
    pkg_check_modules(PKG_XCB QUIET ${pkgConfigModules})

    set(XCB_DEFINITIONS ${PKG_XCB_CFLAGS})
```

But that's the way it was upstream (kde).

Anyway, now I can grab master, and build with the default.nix from https://github.com/corngood/polybar/commit/78f8d36492ffd2f245e22101878bfbd7e6087782.  @mrkgnao Do you want to give it a try?  It's the same as the script you provided except for the src attribute.

I don't know if you're interested, but if you merge that default.nix, anyone who has nix installed should be able to build polybar by simply running `nix-build` in the root, assuming submodules are updated and the working tree is clean.

We could also add polybar to nixpkgs.  @mrkgnao that might be a good exercise for you if you want to make a contribution to https://github.com/NixOS/nixpkgs.  If not, I don't mind doing it.  It should be pretty much the same but with `src = fetchurl { ... }`.

TLDR this should now be fixed by 0a24dc49120ed19e0697f67314f64451744de8bb.

**Comment by jaagr** [comment]
Yeah you are right. I must have read its documentation for an older version of cmake. I guess those modules have some gray hair shafts and cmake tend to be unclear itself at times. Glad the build is working now at least.

**Comment by mrkgnao** [comment]
@jaagr @corngood I'll give it a try in a few. Thanks for helping! :+1:

**Comment by siddharthist** [comment]
@mrkgnao Any progress on adding polybar to nixpkgs?

**Comment by mrkgnao** [comment]
@siddharthist @jaagr I'm sorry, I'm actually working on a polybar-like myself at the moment and don't use polybar at the moment. :/

**Comment by siddharthist** [comment]
@mrkgnao Good to know, I'll put it on my TODO list!

**Comment by NBonaparte** [comment]
Closing, as polybar is already in nixpkgs and the issue appears to be fixed.
